src/language-js/printer-estree.js

```diff
@@ -3863,40 +3863,34 @@ function isSimpleFlowType(node) {
   );
 }
 
-const functionCompositionFunctionNames = new Set([
-  "pipe", // RxJS, Ramda
-  "pipeP", // Ramda
-  "pipeK", // Ramda
-  "compose", // Ramda, Redux
-  "composeFlipped", // Not from any library, but common in Haskell, so supported
-  "composeP", // Ramda
-  "composeK", // Ramda
-  "flow", // Lodash
-  "flowRight", // Lodash
-  "connect", // Redux
-  "createSelector" // Reselect
-]);
-const ordinaryMethodNames = new Set([
-  "connect" // GObject, MongoDB
-]);
-
-function isFunctionCompositionFunction(node) {
-  switch (node.type) {
-    case "OptionalMemberExpression":
-    case "MemberExpression": {
-      return (
-        isFunctionCompositionFunction(node.property) &&
-        !ordinaryMethodNames.has(node.property.name)
-      );
-    }
-    case "Identifier": {
-      return functionCompositionFunctionNames.has(node.name);
-    }
-    case "StringLiteral":
-    case "Literal": {
-      return functionCompositionFunctionNames.has(node.value);
+function looksLikeFunctionCompositionArgs(args) {
+  if (args.length <= 1) {
+    return false;
+  }
+  let functionCount = 0;
+  for (let i = 0; i < args.length; i++) {
+    const arg = args[i];
+    if (
+      arg.type === "FunctionExpression" ||
+      arg.type === "ArrowFunctionExpression"
+    ) {
+      functionCount += 1;
+      if (functionCount > 1) {
+        return true;
+      }
+    } else if (arg.type === "CallExpression") {
+      for (let j = 0; j < arg.arguments.length; j++) {
```

Comment:
Should this recurse?

```js
if (looksLikeFunctionCompositionArgs(arg.arguments)) {
  return true;
}
```

Reply:
I initially thought I would have to write a recursive function as well, but after thinking about it, I decided against it because:
1. I was worried about the additional performance cost of walking the AST while walking the AST
2. checking at a depth of 1 is sufficient for all of the examples of function composition that I’ve found
3. in the course of walking the AST, any nested call expressions with function literals will split on multiple lines, which will automatically trigger the logic for splitting the parent call onto multiple lines.


If you can come up with counter-examples which would be uglier because we don’t check deeper in the tree for function literals I’ll be glad to change it, but honestly I can’t imagine a single example which wouldn’t trigger a line-split, if only because we’d blow the 80 character line budget.

Reply:
Thanks, that makes sense!